packages/eslint-plugin/src/rules/no-misused-promises.ts

```diff
@@ -369,6 +381,66 @@ export default createRule<Options, MessageId>({
       }
     }
 
+    function checkClassLikeOrInterfaceNode(
+      node:
+        | TSESTree.ClassDeclaration
+        | TSESTree.ClassExpression
+        | TSESTree.TSInterfaceDeclaration,
+    ): void {
+      const tsNode = services.esTreeNodeToTSNodeMap.get(node);
+
+      const heritageTypes = getHeritageTypes(checker, tsNode);
+      if (heritageTypes === undefined || heritageTypes.length === 0) {
+        return;
+      }
+
+      for (const nodeMember of tsNode.members) {
+        const memberName = nodeMember.name?.getText();
+        if (memberName === undefined) {
+          continue;
+        }
```

Comment:
[Testing] Either...

* this is a case that can be hit, and so should be covered by tests
* this is a case that can't be hit, and the type system should be told as such so the `if` can be removed

I haven't dug in deep enough to know which it is.

(here and elsewhere)

Reply:
Hey Josh, thanks for the review :)

[Sorry for the long writeup incoming, but here are my thoughts/findings]

I've been looking into this and as far as I can tell, `nodeMember.name` will only ever be `undefined` for call signatures, construct signatures, and index signatures.

### Call signatures
Functions can have multiple call signatures, and types seem to be allowed in TS to have conflicting (overriding?) and duplicate call signatures. Although instantiating the type without TS erroring may require an `any` assertion?
```ts
interface A {
  (): number;
  (): Promise<number>;  // no error
  (): number;  // no error
}
interface B extends A {
  (): boolean; // no error
}
const b: B = () => true;  // Type '() => boolean' is not assignable to type 'B'
const b2: B = (): any => "true";  // no error (other than @typescript-eslint/no-explicit-any)
```
In the context of void- / promise-returning call signatures, I can in theory imagine something like:
```ts
type MySyncCallable = (arg: number) => void;
interface MySometimesAsyncCallable extends MySyncCallable {
  (arg: boolean): Promise<void>;
}
const myCallable: MySometimesAsyncCallable = (arg: number | boolean): any => {
  if (typeof arg === 'number') {
    return;
  }
  return Promise.resolve();
};

myCallable(3); // void
myCallable(true); // Promise<void>
```
Although I'm not sure this is the type of design decision we should be supporting lol / if anyone would even want to do something like this, with the `extends` clause and all. In any case it's a bit unclear to me what the best path is here, and maybe the case of Promise-void-mismatched subtype / heritage type call signatures is niche/unlikely enough that we can just leave it? But please let me know if you have a better grasp of this case and what the best approach is.

### Construct signatures

Constructors can't be async or return void AFAIK so we can just continue to the next `member` if the current one doesn't have a `name` due to being a construct signature.

### Index signatures

Unlike call signatures, TS doesn't seem to allow conflicting index signatures. So this seems more clear to me in terms of expected behavior, and relevant to `no-misused-promises` / `checksVoidReturn` since TS will complain about mismatched index signatures _unless the heritage type index signature is a void-returning function_.

```ts
interface ThisIsNotAllowed {
  [key: string]: () => string;
  [key: `key${string}`]: () => Promise<void>; // () => Promise<void> is not assignable to () => string
}

interface ThisIsAllowedButShouldntBe {
  [key: string]: () => void;
  [key: `key${string}`]: () => Promise<void>;
}
```
For a more full example we probably don't want to allow this:
```ts
type MySyncRecord = Record<`set${string}`, () => void>;
interface MyAsyncRecord extends MySyncRecord {
  someField?: number;
  [key: `set${string}`]: () => Promise<void>;  // this shouldn't be allowed
}

const myAsyncRecord: MyAsyncRecord = {
  setFoo: () => Promise.resolve(),
};

function setFooSync(mySyncRecord: MySyncRecord): void {
  mySyncRecord.setFoo();
}
setFooSync(myAsyncRecord);
```

but currently no errors are shown for the above code (TS or eslint errors).

In the context of `checkClassLikeOrInterfaceNode` when `nodeMember.name === undefined` I haven't actually been able to find a good way to parse the `nodeMember` to get its `keyType` and compare it against the heritage types' index signatures compatible with `keyType`. (I feel like it should be possible in theory, but just not sure what API functions to be using to accomplish that.) But if we wanted we could also just check index signatures all at once separately, something like:

```ts
function checkHeritageTypeIndexSignatures(
  tsNode: ts.ClassLikeDeclaration | ts.InterfaceDeclaration,
  heritageTypes: ts.Type[],
): void {
  const tsNodeType = checker.getTypeAtLocation(tsNode);
  const nodeIndexInfos = checker.getIndexInfosOfType(tsNodeType);
  for (const indexInfo of nodeIndexInfos) {
    if (!anySignatureIsThenableType(checker, tsNode, indexInfo.type)) {
      continue;
    }
    for (const heritageType of heritageTypes) {
      const heritageIndexInfos = checker.getIndexInfosOfType(heritageType);
      for (const heritageIndexInfo of heritageIndexInfos) {
        if (!checker.isTypeAssignableTo(indexInfo.keyType, heritageIndexInfo.keyType)) {
          continue;
        }
        if (!isVoidReturningFunctionType(checker, tsNode, heritageIndexInfo.type)) {
          continue;
        }
        context.report({
          node: services.tsNodeToESTreeNodeMap.get(tsNode),
          messageId: 'voidReturnHeritageTypeMethod',
          data: {
            heritageTypeName: checker.typeToString(heritageType),
          },
        });
      }
    }
  }
}
```

Reply:
> ```ts
> [key: `set${string}`]: () => Promise<void>;
> ```
> ...
> the type of design decision we should be supporting lol

Haha yeah that code is funky. But it _is_ something we need to support. Lint rules generally don't impart judgement on code practices unrelated to their specific purpose. `no-misused-promises` isn't about funky index signatures and someone might legitimately be writing that code. So ideally this rule would support that.

...that being said, I don't know of how to support mapping from `setFoo` to <code>[key: \`set${string}\`]</code> without https://github.com/typescript-eslint/typescript-eslint/discussions/7936. So this might be blocked.

I'll defer to someone else from @typescript-eslint/triage-team if they have time to dig in.

Reply:
> > ```ts
> > [key: `set${string}`]: () => Promise<void>;
> > ```
> > 
> > ...
> > the type of design decision we should be supporting lol
> 
> Haha yeah that code is funky. But it is something we need to support. Lint rules generally don't impart judgement on code practices unrelated to their specific purpose. no-misused-promises isn't about funky index signatures and someone might legitimately be writing that code. So ideally this rule would support that.

Right, yeah good point on not going beyond the specific purpose of the lint rule.

Just to clarify a bit here, it was only the funky call signatures (not index signatures) I wasn't sure we should support—i.e. "should we throw an error if the subtype has any Promise-returning call signatures, when the heritage type only specifies void-returning call signatures?". Technically it could be valid (as in `MySometimesAsyncCallable`), so I think/agree we should probably allow it, and not raise any errors. But the couple potential reasons I was thinking about for raising an error would be:

* I wasn't sure if it was possible to actually use `MySometimesAsyncCallable` without casting to `any` somewhere anyway (going against `@typescript-eslint/no-explicit-any`). But as you point out, our lint rule should probably stay in scope, so that may be irrelevant. (Also I experimented more and it can be done with an `as unknown as _` instead of `any`)
* When we check functions in other contexts for `no-misused-promises` (e.g. [property assignments](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/rules/no-misused-promises.ts#L276-L291)), we already iterate through call signatures in a way that would disallow this funky `MySometimesAsyncCallable` stuff:
  * If the [node has at least one Promise-returning call signature](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/rules/no-misused-promises.ts#L674C1-L682C2), and the [contextual type has at least one void-returning call signature](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/rules/no-misused-promises.ts#L647-L672) and no Promise-returning ones, then we throw an error. So by our current logic, you already wouldn't be able to pass in a `MySometimesAsyncCallable` to a property declared as a `MySyncCallable`, even though it could technically be legal (it could still have the void-returning signatures on it).

Another example showing how we already disallow the funky call signatures:

```ts
type A = () => void;
interface B extends A {
  (asyncArg?: number): Promise<void>;
}
function myWeirdFunction(asyncArg?: number): Promise<void> | void {
  if (asyncArg) {
    return Promise.resolve();
  }
  return;
}

// No TS error, but we report the `no-misused-promises` `voidReturnProperty` error message
const myWeirdA: A = myWeirdFunction;  
```

I think mainly because of this second reason (`no-misused-promises` already disallows the assignment of `MySometimesAsyncCallable` as a `MySyncCallable`) I'm still a bit split on what to do:
1. Ignore call signatures like in `MySometimesAsyncCallable` since "incompatible" call signatures are valid in TS, at least at the type-level before they're actually instantiated
2. Report call signatures like in `MySometimesAsyncCallable` since we already report `no-misused-promises` when a `MySometimesAsyncCallable` is assigned to a `MySyncCallable`

Or a third option, given that we may be blocked on checking index signatures without #7936 (thanks for noticing + calling that out):

3. Ignore all signatures (call or index) and have the main purpose of the new `heritageTypes` option explicitly be to check over named methods only
    * We can just add in a couple test cases with call/index signatures that we expect no errors on, for the code coverage

I think I lean towards 3 on the whole but very open to thoughts. I think we'd already get most of the benefit of the `heritageTypes` option by checking over named members only.

Reply:
Look, I'll be honest with you, we're pretty busy this month (ESLint 9, our v8, other things) and I don't have the time to read through two >500 words on this subject. This might sit for a while.

If you can simplify your findings to something much more scannable that would help, but I understand if you don't have time for that either. 

> “I didn't have time to write a short letter, so I wrote a long one instead.”
> ― Mark Twain

Reply:
Hey Josh, no worries, I understand. First time contributing here so I wasn't sure about the balance between sharing too much/little.

The edge case, and my suggestion, can be simplified to:
* `memberName === undefined` for unnamed methods, which are either: **(i) Call signatures**, **(ii) Index signatures**, or **(iii) Construct signatures**
  1.  **Call signatures**: TypeScript allows conflicting overloaded function signatures (e.g. `() => number` and `() => Promise<number>`), so we shouldn't care about checking mismatches here either (`continue`)
  2. **Construct signatures**: It's not possible for a constructor to be async, so we can also safely ignore these (`continue`)
  3. **Index signatures**: These would be meaningful to check. But to do that properly, we'd want to use `isTypeAssignableTo` (to know which heritage index signatures to check the `member` against). So for this PR, we can just ignore index signatures (`continue`) and leave a comment about implementing index-signature checking once `isTypeAssignableTo` is available.

TLDR suggestion: `continue` past all unnamed methods, specify in the docs that `heritageTypes` only checks named methods, and leave a code comment that we can check index signatures in the future once `isTypeAssignableTo` is available

Reply:
❤️ that sounds good to me, thanks for the summary!

Reply:
Ok yes coming back to this! Confirmed, yes, agreed with your writeups. Let's mark the **iii. Index signatures** case as a followup for post-v8-launch. It'll be a great showcase of the type assignability API!